File: 1143.Longest-Common-Subsequence/dp.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        n, m = len(text1), len(text2)

        # O(m)
        prev = [0] * (m + 1)
        curr = [0] * (m + 1)

        # O(n * m)
        for i in range(1, n + 1):
            for j in range(1, m + 1):
                if text1[i - 1] == text2[j - 1]:
                    curr[j] = 1 + prev[j - 1]
                else:
                    curr[j] = max(prev[j], curr[j - 1])

            prev, curr = curr, prev

        return prev[m]

File: 1143.Longest-Common-Subsequence/dp.rs
impl Solution {
    pub fn longest_common_subsequence(text1: String, text2: String) -> i32 {
        let (n, m) = (text1.len(), text2.len());
        let text1 = text1.as_bytes();
        let text2 = text2.as_bytes();

        // O(m)
        let mut prev = vec![0; m + 1];
        let mut curr = vec![0; m + 1];

        // O(n * m)
        for i in 1..=n {
            for j in 1..=m {
                curr[j] = if text1[i - 1] == text2[j - 1] {
                    1 + prev[j - 1]
                } else {
                    prev[j].max(curr[j - 1])
                };
            }

            (prev, curr) = (curr, prev);
        }

        prev[m]
    }
}
